subsys/bluetooth/audio/bap_broadcast_sink.c

```diff
@@ -157,7 +157,8 @@ static void update_recv_state_big_cleared(const struct bt_bap_broadcast_sink *si
 		return;
 	}
 
-	if (recv_state->encrypt_state == BT_BAP_BIG_ENC_STATE_BCODE_REQ &&
+	if ((recv_state->encrypt_state == BT_BAP_BIG_ENC_STATE_BCODE_REQ ||
+	     recv_state->encrypt_state == BT_BAP_BIG_ENC_STATE_DEC) &&
```

Comment:
From BASS:

> If the BIS is encrypted and the server has the correct encryption key to decrypt the BIS, the server shall write a value of 0x02 (Decrypting) to the BIG_Encryption field of the Broadcast Receive State characteristic.

The `encrypt_state` shall never be `BT_BAP_BIG_ENC_STATE_DEC` if we failed to decrypt the BIG, so this change here looks incorrect to me. 

Are we somewhere setting `BT_BAP_BIG_ENC_STATE_DEC` before we successfully sync to the BIG? If so, then that needs to be changed instead.

Reply:
The test is about to check how the server behaves with a wrong encryption key. In logs it looks like this:
```
[00:00:16.717,620] <dbg> bt_bap_scan_delegator: write_control_point: Assistant setting broadcast code
[00:00:16.717,651] <dbg> bt_bap_scan_delegator: scan_delegator_broadcast_code: Index 0: broadcast code added: 47fc153950449a5d
[00:00:16.717,681] <dbg> bttester_bap: broadcast_code_cb: Broadcast code received for 0x2000c5dc
[00:00:16.888,244] <dbg> bttester_bap: broadcast_sink_bis_sync: 
[00:00:16.888,275] <dbg> bt_bap_broadcast_sink: broadcast_sink_ep_init: ep 0x2000c4c4
[00:00:16.888,824] <dbg> bt_bap_broadcast_sink: broadcast_sink_set_ep_state: ep 0x2000c4c4 id 0x00 idle -> qos-configured
[00:00:17.213,867] <dbg> bt_bap_broadcast_sink: broadcast_sink_iso_connected: stream 0x200033b0
[00:00:17.213,897] <dbg> bt_bap_broadcast_sink: broadcast_sink_set_ep_state: ep 0x2000c4c4 id 0x00 qos-configured -> streaming
[00:00:17.213,928] <dbg> bttester_bap: stream_started: Started stream 0x200033b0
[00:00:17.215,301] <dbg> bt_bap_scan_delegator: bt_bap_scan_delegator_mod_src: Index 0: Source modified: ID 0x00
[00:00:17.215,393] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: Receive State[0]: src ID 0, addr C0:07:E8:2C:B1:C8 (
[00:00:17.215,423] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state:  adv_sid 1, broadcast_id 0xE61038, 
[00:00:17.215,454] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: pa_sync_state 2, encrypt state 2, 
[00:00:17.215,484] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: num_subgroups 1
[00:00:17.215,484] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: 	Subgroup[0]: bis_sync 2 (requested_bis_sync 4294967
[00:00:17.215,667] <dbg> bttester_bap: recv_state_updated_cb: Receive state with ID 0 updated
[00:00:17.218,536] <dbg> bttester_bap: base_recv_cb: 
[00:00:17.218,536] <dbg> bttester_bap: base_recv_cb: Received BASE: broadcast sink 0x2000c39c subgroups 1
[00:00:17.221,618] <dbg> bttester_bap: base_recv_cb: bis_index_bitfield 0x00000002
[00:00:17.224,090] <dbg> bt_bap_broadcast_sink: broadcast_sink_iso_disconnected: stream 0x200033b0 ep 0x2000c4c4 reason 0x3d
[00:00:17.224,121] <dbg> bt_bap_broadcast_sink: broadcast_sink_set_ep_state: ep 0x2000c4c4 id 0x00 streaming -> idle
[00:00:17.224,182] <dbg> bt_bap_scan_delegator: bt_bap_scan_delegator_mod_src: Index 0: Source modified: ID 0x00
[00:00:17.224,273] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: Receive State[0]: src ID 0, addr C0:07:E8:2C:B1:C8 (
[00:00:17.224,304] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state:  adv_sid 1, broadcast_id 0xE61038, 
[00:00:17.224,365] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: pa_sync_state 2, encrypt state 2, 
[00:00:17.224,365] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: num_subgroups 1
[00:00:17.224,395] <dbg> bt_bap_scan_delegator: bt_debug_dump_recv_state: 	Subgroup[0]: bis_sync 0 (requested_bis_sync 4294967
[00:00:17.224,548] <dbg> bttester_bap: recv_state_updated_cb: Receive state with ID 0 updated
[00:00:17.227,172] <dbg> bttester_bap: stream_stopped: Stopped stream 0x200033b0 with reason 0x3D
```
We got `BT_BAP_BIG_ENC_STATE_DEC` state after successful syncing to BIS:
```
[00:00:17.213,867] <dbg> bt_bap_broadcast_sink: broadcast_sink_iso_connected: stream 0x200033b0
```
and since the broadcast code was incorrect as intended in the test case, a little later the server disconnects with an error:
```
[00:00:17.224,090] <dbg> bt_bap_broadcast_sink: broadcast_sink_iso_disconnected: stream 0x200033b0 ep 0x2000c4c4 reason 0x3d
```

Reply:
I see. That's interesting. 

IIRC for the Packetcraft controller, we would get the `HCI_LE_BIG_Sync_Established` event with status = 0x3D (`BT_HCI_ERR_TERM_DUE_TO_MIC_FAIL`), but it seems that for the Zephyr controller (that I assume you are using?) we get the `HCI_LE_BIG_Sync_Established` (which causes the call to `broadcast_sink_iso_connected` followed by a `HCI_LE_BIG_Sync_Lost` event with status = 0x3D (`BT_HCI_ERR_TERM_DUE_TO_MIC_FAIL`)

So basically the BASS implementation follow the state of the BIS in the controller, but I see that may not be good. Ideally we should never enter the `BT_BAP_BIG_ENC_STATE_DEC` state, as that would also trigger a notification to the clients, claiming that we are decrypting the BIG (which we are clearly not doing in this case). 

That being said, since we are guarding this change with `reason == BT_HCI_ERR_TERM_DUE_TO_MIC_FAIL` the change proposed by this PR is harmless. 

@cvinayak I didn't find anything specific in the core spec that says whether the `BT_HCI_ERR_TERM_DUE_TO_MIC_FAIL` should only be sent in the sync lost even, rather than the sync established event, although it is only mentioned in the former HCI event. Do you believe this is the correct and intended behavior, or should the sync fail immediately?